src/main/QueryServer.h

```diff
@@ -0,0 +1,50 @@
+#pragma once
+
+// Copyright 2024 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "lib/httpthreaded/server.hpp"
+
+#include <functional>
+#include <map>
+#include <memory>
+#include <string>
+#include <thread>
+
+namespace stellar
+{
+class SearchableBucketListSnapshot;
+class BucketSnapshotManager;
+
+class QueryServer
+{
+  private:
+    typedef std::function<bool(QueryServer*, std::string const&,
+                               std::string const&, std::string&)>
+        HandlerRoute;
+
+    httpThreaded::server::server mServer;
+
+    std::map<std::thread::id, std::shared_ptr<SearchableBucketListSnapshot>>
```

Comment:
Each thread maintains their own local snapshot object. Originally I just created a new snapshot on demand, but there's a non-trivial cost to creating a new snapshot copy (mostly due to initializing and opening new filestreams). The performance hit isn't noticeable if you're updating your snapshot once every 5 seconds, but creating a new snapshot every 500 us does have an impact.

Reply:
Ah, I see where my confusion came from. `getSearchableBucketListSnapshot` returns a shared ptr, but it actually creates a new object every time (I thought from the interface that it just shares ownership of the same snapshot). I think the code would be more readable if either `getSearchableBucketListSnapshot` returned a const reference (as one might expect from a getter) and then the code here created shared ptrs explicitly with `make_shared`, or rename `getSearchableBucketListSnapshot` to `copySearchableBucketListSnapshot` or something like that (basically hint at creating a new copy). I would prefer option 1 as it makes client uses more explicit (easy to distinguish allocating a new objects from read-only access if we ever need it).

Reply:
I think I'll go with option 2. Reason being we want to avoid race conditions as much as possible. If we return a `const reference` and leave it up to the caller to call `make_shared`, it's possible different threads could race if callers have a bug. I'll change the name, and I'll also change the return type to a `unique_ptr` to better clarify that each caller should maintain their own copy of the snapshot.

Reply:
If you're concerned about race conditions, shouldn't you just add `releaseAssert(threadIsMain());` to the getter? Basically if a method is not thread-safe, you should either not call it from different thread, or you should add synchronization.

Reply:
Yeah that's the issue with returning a const reference. The current interface where a copy is returned is thread safe. Any type of const reference return where the caller would need to instantiate a copy is no longer thread safe.

Reply:
If `this` can be mutated by a different thread, then any access to it can cause race condition, right? It doesn't really matter what do you do with `this` in this method, you must either access it from the main thread only, or you need to synchronize access to it.

Reply:
Sorry, I think I mispoke. `getSearchableBucketListSnapshot` is thread safe, as it just calls the copy constructor:
```
SearchableBucketListSnapshot::SearchableBucketListSnapshot(
    BucketSnapshotManager const& snapshotManager)
    : mSnapshotManager(snapshotManager), mHistoricalSnapshots()
{
    // Initialize snapshot from SnapshotManager
    mSnapshotManager.maybeUpdateSnapshot(mSnapshot, mHistoricalSnapshots);
}
```

where `maybeUpdateSnapshot` acquires the lock. The function will be thread safe regardless. The object returned is only thread safe if we return a new copy. If we return const reference, the returned object would not be thread safe even though the getter function itself is.